package roomescape.testRepository;

import java.util.ArrayList;
import java.util.List;
import java.util.Optional;
import roomescape.domain.ReservationTime;
import roomescape.repository.TimeRepository;

public class FakeTimeRepository implements TimeRepository {

    private final List<ReservationTime> reservationTimes = new ArrayList<>();

    private Long index = 0L;

    @Override
    public Long save(ReservationTime reservationTime) {
        ReservationTime timeWithId = ReservationTime.withId(++index, reservationTime);
        reservationTimes.add(timeWithId);
        return index;
    }

    @Override
    public List<ReservationTime> findAll() {
        return reservationTimes;
    }

    @Override
    public Optional<ReservationTime> findById(Long id) {
        return reservationTimes.stream()
                .filter(time -> time.getId().equals(id))
                .findFirst();
    }

    @Override
    public void deleteById(Long id) {
        ReservationTime reservationTime = findById(id)
                .orElseThrow(() -> new IllegalArgumentException("FakeTimeRepository: 삭제하려는 id 없음"));
        reservationTimes.remove(reservationTime);
    }
}
